magento/pwa-studio
# PWA-1845: Add to Cart on Category Listing - Out of stock is visible

- Updated poduct detailed page to correctly show when product is out of stock for simple and configurable products

## Description
Note: The general category page implementation is already in place, and thus the only development needed was for the PDP pages to ensure the CTA behaves as expected for out of stock situations.

As a Shopper I want to add products to the cart when I browse the category listing page *so that* I don't need to go to the product detail page in case I'm already familiar with a product and know what I'm buying

*Preconditions*
 1. Store with 2 store views exists 
 2. Simple and configurable products created with localization for store view 1 and store view 2
 3. Simple product 1 is in stock, simple product 2 is out of stock 
 4. Configurable product 1 has stock options on its variations 
 5. Configurable product 2 has all the variations out of stock 
 6. Bundle and virtual products created 
7. Stores > Settings > Configuration > Catalog > Inventory>Stock Options>Display Out of Stock Products id set to Yes

*Acceptance Criteria*
 * User opens category where simple 1 and simple 2 belong and sees both simple 1 and 2 in the listing
 * User sees CTA that enables adding simple 1 to the cart
 * User adds simple 1 to the cart and sees the mini cart updated
 * User sees the indication of simple 2 is out of stock
 * User has no option to add simple 2 into the cart
 * When user navigates to the simple 2 PDP, they see the stock status and CTA not available
 * User opens a category with configurable 1 and configurable 2
 * User sees the CTA on configurable 1 that navigates to the PDP
 * User sees the indication that configurable 2 is out of stock
 * User has no option to add configurable 2 into the cart
 * User sees no option to add bundle and virtual products to the cart
Adding out of stock button CTA for simple and configurable products on the product detail pages.

## Related Issue







Closes https://jira.corp.magento.com/browse/PWA-1845

## Acceptance





### Verification Stakeholders



### Specification




## Verification Steps

* Verify user goes to Store View 1 to category where Simple 1 and simple 2 products belong : 
 1. Verify Add To Cart button exists for Simple 1 product
 2. verify clicking on Add to Cart button >> adds products to mini cart (Check for default empty cart, guest user and cart which already had products, Registered user)
 1. verify clicking again on Add to Cart still adds same product to cart and updates mini cart
 2. verify Out of Stock button exists for Simple 2 product, and cannot add to cart
 3. verify Simple 2 PDP page displays the stock status, and CTA does not exist
 * When user goes to Store View 1 to category where Configurable 1 and Configurable 2 products belong : 
 1. Verify Add To Cart button exists for Config 1 product
 2. verify clicking on Add to Cart button >> user lands on PDP pages (Check guest user and Registered user) and make sure user is able to add product to cart from PDP
 3. verify Config 2 product display out of stock indication, and cannot add to cart
 * Verify Bundle and Virtual products cannot be added to the cart
 * Verify Button display and Add to Cart interaction on Page 2 (any page other than default page landing page) of category list as well.
 * Verify above scenarios by switching to Store View 2
 * Verify Product listing on Search page does not have these changes.
 * Verify display and interaction of Add to Cart still works when Filter and Sorting is applied 
 * Verify button display on Chrome, Safari, FF and Android/iOS device
 * Run MFTF and Cypress


#### Test scenario(s) for direct fix/feature





#### Test scenario(s) for any existing impacted features/areas





#### Test scenario(s) for any Magento Backend Supported Configurations





#### Is Browser/Device testing needed?





#### Any ad-hoc/edge case scenarios that need to be considered?




## Screenshots / Screen Captures (if appropriate)
Simple Product Out of Stock
![simple product out of stock](https://user-images.githubusercontent.com/9815466/131380210-c265dc04-9b71-4fd4-a6b1-fe7262c86ed3.png)

Configurable Product Initialized
![config product initialized](https://user-images.githubusercontent.com/9815466/131380257-d924924d-f98f-42d5-ae03-38490d7e6265.png)

In Stock Configurable Product Option
![in stock config product option](https://user-images.githubusercontent.com/9815466/131380337-b18b1f00-6d85-49a2-8b16-894788a32a84.png)

Out of Stock Configurable Product Option
![out of stock config product option](https://user-images.githubusercontent.com/9815466/131380355-f8b8ef60-a3e1-4363-9984-fd951fd746d2.png)

## Breaking Changes (if any)





## Checklist



-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.
